File: Compiler/Parser/Trait.py
import Compiler.Syntax.Trait as Trait
import Compiler.Parser.Function as Function
import Compiler.Parser.Type as Type

def parseTrait(parser, module_name):
    parser.expect("trait")
    trait = Trait.Trait()
    if parser.peek("leftbracket"):
        trait.generics = Type.parseGenericDeclaration(parser)
    trait.name = parser.parseTypeName()
    parser.expect("leftbracket")
    dependentParams = False
    while True:
        param = parser.parseTypeName()
        trait.generic_parameters.append(param)
        if parser.peek("rightbracket"):
            break
        if parser.peek("greaterthan"):
            parser.expect("greaterthan")
            dependentParams = True
            break
        parser.expect("comma")
    if dependentParams:
        while True:
            param = parser.parseTypeName()
            trait.dependent_parameters.append(param)
            if parser.peek("rightbracket"):
                break
    parser.expect("rightbracket")
    if parser.peek("leftcurly"):
        parser.expect("leftcurly")
        while True:
            if parser.peek("fn"):
                fn = Function.parseFunction(parser, module_name)
                if fn.body is None:
                    trait.method_declarations.append(fn)
                else:
                    trait.methods.append(fn)
            else:
                break
        parser.expect("rightcurly")
    return trait

def parseInstance(parser, module_name):
    parser.expect("instance")
    instance = Trait.Instance()
    if parser.peek("leftbracket"):
        instance.generics = Type.parseGenericDeclaration(parser)
    instance.type = Type.parseType(parser)
    if parser.peek("leftcurly"):
        parser.expect("leftcurly")
        while True:
            if parser.peek("fn"):
                fn = Function.parseFunction(parser, module_name)
                if fn.body is None:
                    instance.declarations.append(fn)
                else:
                    instance.methods.append(fn)
            else:
                break
        parser.expect("rightcurly")
    return instance

File: sikors/src/siko/parser/Trait.rs
use std::mem;

use crate::siko::syntax::Trait::{Instance, Trait};

use super::{
    Function::FunctionParser,
    Parser::Parser,
    Token::{BracketKind, KeywordKind, MiscKind, OperatorKind, TokenKind},
    Type::TypeParser,
};

pub trait TraitParser {
    fn parseTrait(&mut self) -> Trait;
    fn parseInstance(&mut self) -> Instance;
}

impl TraitParser for Parser {
    fn parseTrait(&mut self) -> Trait {
        self.expect(TokenKind::Keyword(KeywordKind::Trait));
        let typeParams = if self.check(TokenKind::LeftBracket(BracketKind::Square)) {
            Some(self.parseTypeParameterDeclaration())
        } else {
            None
        };
        let name = self.parseTypeIdentifier();
        let mut params = Vec::new();
        let mut depParams = Vec::new();
        let mut isDep = false;
        self.expect(TokenKind::LeftBracket(BracketKind::Square));
        loop {
            let param = self.parseTypeIdentifier();
            if isDep {
                depParams.push(param);
            } else {
                params.push(param);
            }
            match self.peek() {
                TokenKind::RightBracket(BracketKind::Square) => break,
                TokenKind::Misc(MiscKind::Comma) => {
                    self.expect(TokenKind::Misc(MiscKind::Comma));
                }
                TokenKind::Op(OperatorKind::GreaterThan) => {
                    if isDep {
                        self.reportError2(", or ]", TokenKind::Op(OperatorKind::GreaterThan));
                    }
                    isDep = true;
                    self.expect(TokenKind::Op(OperatorKind::GreaterThan));
                }
                kind => self.reportError2(", or ]", kind),
            }
        }
        self.expect(TokenKind::RightBracket(BracketKind::Square));
        let mut members = Vec::new();
        if self.check(TokenKind::LeftBracket(BracketKind::Curly)) {
            self.expect(TokenKind::LeftBracket(BracketKind::Curly));
            while !self.check(TokenKind::RightBracket(BracketKind::Curly)) {
                let function = self.parseFunction();
                members.push(function);
            }
            self.expect(TokenKind::RightBracket(BracketKind::Curly));
        }
        Trait {
            name: name,
            typeParams: typeParams,
            members: members,
        }
    }

    fn parseInstance(&mut self) -> Instance {
        self.expect(TokenKind::Keyword(KeywordKind::Instance));
        let typeParams = if self.check(TokenKind::LeftBracket(BracketKind::Square)) {
            Some(self.parseTypeParameterDeclaration())
        } else {
            None
        };
        let ty = self.parseType();
        let mut members = Vec::new();
        if self.check(TokenKind::LeftBracket(BracketKind::Curly)) {
            self.expect(TokenKind::LeftBracket(BracketKind::Curly));
            while !self.check(TokenKind::RightBracket(BracketKind::Curly)) {
                let function = self.parseFunction();
                members.push(function);
            }
            self.expect(TokenKind::RightBracket(BracketKind::Curly));
        }
        Instance {}
    }
}
